compiler/noirc_evaluator/src/ssa/opt/die.rs

```diff
@@ -146,15 +150,21 @@ impl Context {
             self.track_inc_rcs_to_remove(
                 *instruction_id,
                 function,
+                &mut rcs_with_possible_pairs,
+                &mut rc_pairs_to_remove,
                 &mut inc_rcs,
-                &mut inc_rcs_to_remove,
+                &mut borrowed_arrays,
             );
         }
 
-        for id in inc_rcs_to_remove {
-            self.instructions_to_remove.insert(id);
+        for inc_rc_value in inc_rcs.keys() {
+            if !borrowed_arrays.contains(inc_rc_value) {
+                self.instructions_to_remove.extend(&inc_rcs[inc_rc_value]);
+            }
         }
```

Comment:
```suggestion
        let unborrowed_arrays = inc_rcs.keys().filter(|id| !borrowed_arrays.contains(id));
        self.instructions_to_remove.extend(unborrowed_arrays);
```

nit

Reply:
`inc_rcs[inc_rc_value]` actually returns `HashSet` of instructions so I had to change this suggestion to be a `filter_map` with `flatten`. It ends up being longer than the for-loop version, but it is more functional.